win32ss/user/winsrv/consrv/frontends/gui/text.c

```diff
@@ -501,8 +501,136 @@ GuiPaintTextModeBufferCJK(PTEXTMODE_SCREEN_BUFFER Buffer,
+    WORD Attribute;
+    ULONG CursorX, CursorY, CursorHeight;
+    ULONG Line, Column;
+    HBRUSH CursorBrush, OldBrush;
+    HFONT OldFont, NewFont;
+    BOOLEAN IsUnderline;
+    PCHAR_INFO Ptr;
+
+    ConioInitLongRect(rcFramebuffer, 0, 0, 0, 0);
+
+    if (Buffer->Buffer == NULL)
+        return;
+
+    if (!ConDrvValidateConsoleUnsafe((PCONSOLE)Console, CONSOLE_RUNNING, TRUE))
+        return;
+
+    ConioInitLongRect(rcFramebuffer,
+                      Buffer->ViewOrigin.Y * GuiData->CharHeight + rcView->top,
+                      Buffer->ViewOrigin.X * GuiData->CharWidth  + rcView->left,
+                      Buffer->ViewOrigin.Y * GuiData->CharHeight + rcView->bottom,
+                      Buffer->ViewOrigin.X * GuiData->CharWidth  + rcView->right);
+
+    LeftColumn  = rcFramebuffer->left  / GuiData->CharWidth;
+    RightColumn = rcFramebuffer->right / GuiData->CharWidth;
+    if (RightColumn >= (ULONG)Buffer->ScreenBufferSize.X)
+        RightColumn = Buffer->ScreenBufferSize.X - 1;
+
+    TopLine    = rcFramebuffer->top    / GuiData->CharHeight;
+    BottomLine = rcFramebuffer->bottom / GuiData->CharHeight;
+    if (BottomLine >= (ULONG)Buffer->ScreenBufferSize.Y)
+        BottomLine = Buffer->ScreenBufferSize.Y - 1;
+
+    Attribute = ConioCoordToPointer(Buffer, LeftColumn, TopLine)->Attributes;
+    SetTextColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, TextAttribFromAttrib(Attribute)));
+    SetBkColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, BkgdAttribFromAttrib(Attribute)));
+    IsUnderline = !!(Attribute & COMMON_LVB_UNDERSCORE);
+
+    /* Select the new font */
+    NewFont = GuiData->Font[IsUnderline ? FONT_BOLD : FONT_NORMAL];
+    OldFont = SelectObject(GuiData->hMemDC, NewFont);
+
+    for (Line = TopLine; Line <= BottomLine; Line++)
+    {
+        for (Column = LeftColumn; Column <= RightColumn; Column++)
+        {
+            Ptr = ConioCoordToPointer(Buffer, Column, Line);
+            Attribute = Ptr->Attributes;
+            SetTextColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, TextAttribFromAttrib(Attribute)));
+            SetBkColor(GuiData->hMemDC, PaletteRGBFromAttrib(Console, BkgdAttribFromAttrib(Attribute)));
+
+            /* Change underline state if needed */
+            if (!!(Attribute & COMMON_LVB_UNDERSCORE) != IsUnderline)
+            {
+                IsUnderline = !!(Attribute & COMMON_LVB_UNDERSCORE);
+
+                /* Select the new font */
+                NewFont = GuiData->Font[IsUnderline ? FONT_BOLD : FONT_NORMAL];
+                SelectObject(GuiData->hMemDC, NewFont);
+            }
+
+            if (Attribute & COMMON_LVB_TRAILING_BYTE)
+                continue;
+
+            TextOutW(GuiData->hMemDC,
+                     Column * GuiData->CharWidth,
+                     Line * GuiData->CharHeight,
+                     &Ptr->Char.UnicodeChar, 1);
+        }
+    }
+
+    /* Restore the old font */
```

Comment:
Now everything downwards can be again shared between a common implementation of `GuiPaintTextModeBuffer()` and there is no problem in checking then whether the console is in CJK mode in order to display the caret wide (depending on the character attributes) or not.